### PR TITLE
fix issue 20653 .  Short circuit manifest constant evaluation

### PR DESCRIPTION
**Pull request by benjones**
The fix is adding the SCOPE.condition flag when doing CTFE to infer the VarDecl type in dsymbolsem.  

The rest of the changes are fixes to the debug logging.  I can remove them if they're distracting

### CODE REVIEW

**Comment by WalterBright** [comment]
Please explain the point of this. Is it a bug fix? An enchancement?

**Comment by thewilsonator** [comment]
>  Is it a bug fix?

>fix issue 20653

Dunno where the bot is though.

**Comment by thewilsonator** [comment]
>The rest of the changes are fixes to the debug logging. I can remove them if they're distracting

Can you split them into a different commit?

**Comment by benjones** [comment]
I started this as a draft PR, so maybe @dlang-bot didn't notice when I dropped the draft.

I'll split into 2 commits.  I'll also add a test with &&

**Review by thewilsonator — APPROVED** [review]

**Comment by Geod24** [comment]
Could you add a test with a template instantiation ?
Something like:

```
T myFunc() { return T.init; }
enum Foo(T) = is(T == int) || myFunc!T();

void main ()
{
    assert(myFunc!int() == 0);
}
```

The idea is to check that `myFunc!int` is properly emitted and does not cause linking error.

**Comment by TurkeyMan** [comment]
> Dunno where the bot is though.

I mean, cheeky bugger's probably self-isolating!

**Comment by WalterBright** [comment]
> cheeky bugger's

I can't let you Aussies have a monopoly on cool phrases. It's not fair!

**Comment by TurkeyMan** [comment]
😛

**Comment by MoonlightSentinel** [comment]
> Could you add a test with a template instantiation ?
> Something like:
> 
> ```d
> T myFunc() { return T.init; }
> enum Foo(T) = is(T == int) || myFunc!T();
> 
> void main ()
> {
>     assert(myFunc!int() == 0);
> }
> ```
> 
> The idea is to check that `myFunc!int` is properly emitted and does not cause linking error.

Could also add this to the `compilable` test and force linking with

```
LINK:
```

File: src/dmd/dsymbolsem.d [hunk 1]
```diff
@@ -805,9 +806,11 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
             // so mark the scope as ctfe if required
             bool needctfe = (dsym.storage_class & (STC.manifest | STC.static_)) != 0;
             if (needctfe)
+            {
+                sc.flags |= SCOPE.condition;
```
Comment (ntrel) [inline src/dmd/dsymbolsem.d:810]: @benjones This breaks the error for `is` *Identifier* forms used outside of static if/assert:
https://issues.dlang.org/show_bug.cgi?id=23827

I tried making a fix but it doesn't pass the test here.

Comment (benjones) [reply src/dmd/dsymbolsem.d:810]: I saw your new PR.  It's been so long, I don't remember how I came to think this was the right fix and probably won't have time to look closely in the next couple of weeks.

**Pull request merged**